### ISSUE TITLE
Make it easier to write value transformers

### ISSUE THREAD

**Issue by bgrant0607** [GoogleContainerTools/kpt#3155]
Value transformers are critical to the power of kustomize and kpt, which really differentiate the approaches from monolithic configuration generators. We're going to need to build a lot more of them, and we need to make it as easy as possible to write such transformers, including guides (documentation, examples, videos, etc.), so that users don't fall back to less Config-as-Data-friendly approaches. 

We’ve implemented several approaches to value propagation.

kustomize:
1. Kustomize transformer plugins ([guide](https://kubectl.docs.kubernetes.io/guides/extending_kustomize/), [example](https://blog.scottlowe.org/2020/03/13/configuring-kustomize-transformers-for-cluster-api/))
2. Kustomize vars (templates, more or less)
3. Kustomize replacement transformer ([example use](https://youtu.be/Xoh_OpLoVtI?t=871))

kpt:
4. Pre-1.0 kpt setters and substitutions
5. Kpt 1.0 setters
6. Catalog specific-purpose transformers, such as set-namespace

The appealing aspect of (1) is that configuration authors and consumers only need to specify the values to propagate, not where to propagate them to. Where they need to be propagated to is consistent for a given type of value and a given resource type, or set of resource types for horizontal transforms. 

Transforms that apply to many resource types  or all types need to be written in a different manner than those that apply to one or two specific types. Some transformers are a mix of the two styles.

Perhaps we could even make a declarative value propagator function, like a more powerful replacement transformer with configuration that's reusable across packages.

* Replicas example: [plugin](https://github.com/kubernetes-sigs/kustomize/blob/master/plugin/builtin/replicacounttransformer/ReplicaCountTransformer.go), [filter](https://github.com/kubernetes-sigs/kustomize/blob/master/api/filters/replicacount/replicacount.go), [fields](https://github.com/kubernetes-sigs/kustomize/blob/master/api/konfig/builtinpluginconsts/replicas.go)
* Images example: [plugin](https://github.com/kubernetes-sigs/kustomize/blob/master/plugin/builtin/imagetagtransformer/ImageTagTransformer.go), [filter](https://github.com/kubernetes-sigs/kustomize/blob/master/api/filters/imagetag/imagetag.go), [fields](https://github.com/kubernetes-sigs/kustomize/blob/master/api/konfig/builtinpluginconsts/images.go), [example](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/transformerconfigs/README.md#images-transformer) (it’s a problem that the kinds are not specified)

The transformer functions require understanding the schemas of applicable resource types, at least the field paths of interest. Eventually we should ensure that types of interest publish sufficient information attached to the OpenAPI schemas and that kpt can read it, or that there's a standard way to patch OpenAPI schemas with such information, but for now we should use an [approach similar to kustomize](https://github.com/kubernetes-sigs/kustomize/blob/master/api/filters/imagetag/imagetag.go), and [enumerate relevant fields](https://github.com/kubernetes-sigs/kustomize/tree/master/api/konfig/builtinpluginconsts) in their varying representations, and also support user-specified paths (again, in a way that's reusable across packages).  

We will eventually want a better mechanism for defining value types and mapping resource attributes (by value type and by [resource type](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/transformerconfigs/crd/README.md)) to them so that users can easily add support for [additional resource types](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/transformerconfigs/README.md) ([skaffold image replacement example](https://github.com/GoogleContainerTools/skaffold/issues/4081), [ClusterAPI example](https://blog.scottlowe.org/2020/03/13/configuring-kustomize-transformers-for-cluster-api/)). As one example, kustomize’s object reference and label selector transformations don’t work for arbitrary resource types out of the box because it only has data for built-in Kubernetes types.

To simplify targeting/filtering expressions, we should explore leveraging the package structure (nested subpackages), but we’ll need to look more at example packages to validate that. Hierarchical dictionaries of facts are common in configuration systems: [hiera](https://puppet.com/docs/puppet/7/hiera.html), [facter](https://puppet.com/docs/puppet/7/facter.html).

In addition to performing the replacements, we’ll also want to be able to identify replacement locations, such as for documentation and introspection.

Some example functions that apply to non-built-in types we could try to write include #2562, #2658, #2677, #1400.

**Comment by mikebz** [comment]
I think the problem with even the simplest SDK is that I have to build the image and publish it.  I would advocate that for scenarios like that and internal functions the easiest thing to do is to build in Starlark and run it without a container.  kpt simple transformer can not be much harder than `bash` + `sed`.

Templating of Helm values.yaml does not make sense because you can just put the right value in the values.yaml file and update the template to put it in the right places.  It's not much more difficult than `sed`.

**Comment by yuwenma** [comment]
> I think the problem with even the simplest SDK is that I have to build the image and publish it. I would advocate that for scenarios like that and internal functions the easiest thing to do is to build in Starlark and run it without a container. 


@mikebz That's right. We want users to use starlark for simple scenarios and go SDK for more complicated cases. Though we have a pile of KRM function tasks and I don't think starlark is more mature than go SDK to recommend. Another reason is that the above use case is more like a set-image transformation, which assume might need to import some logic from the set-image, which is in Go.

> kpt simple transformer can not be much harder than `bash` + `sed`.
I like this description. It can be the standards to evaluate our starlark SDK.

**Comment by bgrant0607** [comment]
One original motivation for templating values.yaml was because the files can get huge. The ability to merge multiple values.yaml files mitigated that issue.

However, if a value is in an environment variable, then automatically getting it into values.yaml, such as in a CI/CD pipeline, has the same challenge we're discussing here with kpt.

I understand the Starlark SDK is less mature.

kpt still supports exec functions, doesn't it? So I think even with Go a container isn't required. 

We could also make it easy to build CLIs that source, run functions, then sink, without needing to invoke the kpt CLI.

But, yes, yq or envsubst could work in this case, especially if the configuration could live outside the Kptfile.

That reminds me that we could use a ConfigMap generator: #3119.

**Comment by droot** [comment]
I think it's worth sharing the recipe of propagating values from environment variable to any KRM fields. For example, [this use-case](https://github.com/GoogleContainerTools/kpt/issues/3155#issuecomment-1151527968) can be addressed by invoking `search-replace` function imperatively. I am assuming that substituting repo name in the deployment spec needs to be done once.

Here are the steps involved:

Assuming following `Kptfile` in the `app-base` package:
```
# Kptfile in app-base package
apiVersion: kpt.dev/v1
kind: Kptfile
metadata:
  name: app-base
  annotations:
    config.kubernetes.io/local-config: "true"
info:
  description: kpt package for provisioning a deployment
pipeline:
  mutators:
  - image: gcr.io/kpt-fn/set-image:v0.1.1
    configMap:
      name: bla/a
      newName: newblabla/a
      newTag: latest
  - image: gcr.io/kpt-fn/set-image:v0.1.1
    configMap:
      name: bla/b
      newName: newblabla/b
      newTag: latest
``` 

Assume that`app-base` also contains deployment.yaml:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: frontend
  name: frontend
spec:
  replicas: 1
  selector:
    matchLabels:
      app: frontend
  strategy: {}
  template:
    metadata:
      labels:
        app: frontend
    spec:
      containers:
      - name: frontend
        image: bla/a
      - name: backend
        image: bla/b
```

Let's create deployable instance from the app-base blueprint:

```sh
$ kpt pkg get <app-base> --for-deployment

# Assuming REPO_NAME environment variable contains the name `app1`
$ kpt fn eval -i search-replace:v0.2.0 -- by-value="newblabla/a" put-value="$REPO_NAME/a"
[RUNNING] "gcr.io/kpt-fn/search-replace:v0.2.0"
[PASS] "gcr.io/kpt-fn/search-replace:v0.2.0" in 1.1s
  Results:
    [info] pipeline.mutators[0].configMap.newName: Mutated field value to "app1/a"

$ kpt fn eval -i search-replace:v0.2.0 -- by-value="newblabla/b" put-value="$REPO_NAME/b"
[RUNNING] "gcr.io/kpt-fn/search-replace:v0.2.0"
[PASS] "gcr.io/kpt-fn/search-replace:v0.2.0" in 900ms
  Results:
    [info] pipeline.mutators[1].configMap.newName: Mutated field value to "app1/b"

# Now let's run render on it.
$ kpt fn render
Package "app-base":
[RUNNING] "gcr.io/kpt-fn/set-image:v0.1.1"
[PASS] "gcr.io/kpt-fn/set-image:v0.1.1" in 900ms
  Results:
    [info] apps/v1/Deployment/frontend spec.template.spec.containers.image: set image from bla/a to app1/a:latest
[RUNNING] "gcr.io/kpt-fn/set-image:v0.1.1"
[PASS] "gcr.io/kpt-fn/set-image:v0.1.1" in 700ms
  Results:
    [info] apps/v1/Deployment/frontend spec.template.spec.containers.image: set image from bla/b to app1/b:latest

Successfully executed 2 function(s) in 1 package(s).

```

I just realized `set-image` isn't very useful for declarative workflow because now images field have new values, `kpt fn render` re-runs won't do anything.
